### ISSUE TITLE
k8s-externalipcontroller like features

### ISSUE THREAD

**Issue by mleklund** [google/metallb#187]
**Is this a bug report or a feature request?**:

feature request

With regard to ARP operation, I wondered if you had seen k8s-externalipcontroller.
https://github.com/Mirantis/k8s-externalipcontroller

A few of nice things their controller does:
1) adds ips via netutils so that they appear in the networking config for server via `ip addr sh`
2) apply ip address to multiple nodes, somewhat distributing the load.
3) Use custom resources for claim pools

**Comment by danderson** [comment]
I wasn't aware of externalipcontroller, thanks for the reference!

Now, some thoughts... Sorry for the long reply, I want to explain in detail.

1. Adding IPs to interfaces: MetalLB does not do this, by design.

I agree that it would be nice to have standard linux tools give useful data about load-balanced IPs. However, adding IPs to interfaces comes with some significant downsides:

- There is no foolproof way to clean up interface addresses. This means we give up the current desirable behavior of MetalLB: if you tear down MetalLB, all traces of what it does to the network vanish immediately.
- It increases the risk of interoperability problems with CNI plugins, and with kube-proxy. These two components co-own most of the node network stack, and the more you touch it, the more you risk breaking things.
  - A perfect example is kube-proxy's new IPVS mode: in this mode, kube-proxy _does_ add LB IPs to a special interface to make IPVS work. If MetalLB did it as well, we would end up with conflicts, and we would be forced to detect what "flavor" of kube-proxy is being used (and react correctly if user changes the mode while the machine is running - which is 100% allowed).
  - Another example is that kube-proxy configures a bunch of netfilter rules to allow service traffic to flow in the way that it's configured. However, adding those IPs to an interface changes the packet flow in netfilter (instead of going through the FORWARD tables, the traffic goes through the INPUT tables). Depending on what kube-proxy does, this could completely break the traffic processing. Or it might be fine... But now we are depending on the specific version and configuration of kube-proxy, and I don't have the resources to do enough interop testing to guarantee that it's safe.

2. Apply IPs to multiple nodes: no, this makes the system hard to debug.

I'm assuming you're suggesting this for ARP mode, because in BGP mode we already have load distribution. I agree, it would be nice if ARP mode could distribute load!

What you're suggesting here is essentially: every machine in the cluster should respond to ARP requests, and whichever one got lucky and responded first "wins". In theory, if timings are slightly different each time, different clients might cache different nodes.

The problem is that this is not reliable in all cases, and makes it nearly impossible to debug.

It's not reliable, because it assumes that a client will only ask for the server MAC once, and then keep using it for one entire "session" (TCP connection, HTTP request, whatever). This is mostly true, but only for OSes that use upper layer protocol signals to refresh the ARP cache (i.e. each time TCP processes an incoming packet, it tells ARP "BTW, IP x contacted me with MAC y").

I can't guarantee that all clients do this, because AFAIK it's not required by any standards, it's just an optimization. And if a client doesn't implement this (e.g. some IoT thing with a tiny network stack), the result is that connections break unexpectedly when the client sends a new ARP request and gets a different MAC in response.

The non-deterministic behavior makes it very hard to debug. A client complains that they get connection refused from a service, but it works for you. You try from many different places, and it still works, because by pure luck the bad node is repeatedly "losing" the ARP race... But your client has cached the bad MAC, and each time they retry and get a connection refused, the ARP cache gets refreshed by TCP and doesn't expire.

And all this buys you is non-deterministic, uneven load distribution. You might get lucky and get acceptable distribution, or maybe the switch you're plugged into has a subtle timing weirdness where port number 4 is a tiny bit faster to forward, and so one machine consistently always wins the race... So you have no way of predicting what your "real" capacity is. Again, hard to debug because when the luck of the timing changes, suddenly your effective capacity might drop from 10 machines to 1, with no obvious reason.

My general feeling is that, if you care about load _balancing_, it sounds like it's a production cluster, and you should really be using BGP (or helping to implement RIP, which can also distribute load). If you just care about "I want my external services to just work", ARP/NDP is fantastic for that, but the tradeoff is that you don't get the nice features from more complex protocols.

3. Use custom resources for claim pools

I'm not sure I understand what you mean here... Looking at the demo for k8s-externalipcontroller, it looks like it's using a custom resource type to record "I should make this IP externally reachable." MetalLB should definitely not use this, we're using the `status.loadBalancer` field of service objects, which is the standard extension point in k8s for LB implementations.

But, maybe you mean we should use a similar mechanism to define address pools, instead of using a ConfigMap? That's... possible, and actually might solve some other problems we need to solve, like #161 (need a place to put an audit log for allocations, but we don't have a good place for that today... Custom resources could be a solution). That's quite a lot of work, but if this is what you're suggesting, we can file a bug for that to track it!


Again, sorry for the huge reply. I realize I sound very negative, but I just think k8s-externalipcontroller and MetalLB have some incompatible goals for some of this stuff, and I wanted to explain in detail why I think this, instead of just saying "no!" Hopefully that's helpful :)

**Comment by danderson** [comment]
Correction: k8s-externalipcontroller _does_ do leader election per IP, and so distributes the load by sharding by IP. Apologies for being completely wrong on point 2.

Okay, so _that_ is somewhat feasible. I'm unconvinced that it's worth it, because honestly if traffic is big enough that you're trying to shard, you should probably still be switching to BGP... But it's tractable, at least.

Upside: load distribution. Downside: have to run 1 leader election per IP, which is pretty heavy on the k8s control plane. Still not entirely convinced on the benefits, but it's definitely worth considering.

**Comment by mleklund** [comment]
All of your points are fair.

Clarifying 1, I mostly brought up nettools, visible ip because it is nice to be able to see what holds the ip without talking to the switch, though I concede at the end of the day the switch is the source of truth.  Also, I with respect to arp failover times, it is what keepalived is doing.  I wondered if that gave any kind of advantage to arp failover times, as stated in another ticket, I have always seen near instantaneous failover with keepalived.

Clarifying 3, custom I had in mind when commenting on #161.  

And to be clear, I am very impressed with how well thought out the design of metallb is and pointing to k8s-externalipcontroller was mostly "here is a thing that does some of the same stuff you are doing, but in a different way".  As a developer I love to look at other peoples work, especially if it is in the same domain as my own.  I have never cornered the market on bright ideas, far to often it is quite the opposite.

The ticket has served it's purpose, so feel free to close whenever you are ready.

**Comment by danderson** [comment]
Thanks. Again, I know I sounded very negative, but I really do appreciate the opportunity to see how others have done things, and if we can use some of those ideas.

I still need to file bugs for points 2+3, there's possible features in there. The outcome down the road might be that I close those bugs as "no, we won't implement that", but until I can think about it seriously, I want to keep track of the suggestions.

I'll close this bug once I've filed the related ones.
